runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/BatchFlinkExecutableStageContext.java

```diff
@@ -69,32 +75,132 @@ public StateRequestHandler getStateRequestHandler(
   }
 
   @Override
-  protected void finalize() throws Exception {
+  public void close() throws Exception {
     jobBundleFactory.close();
   }
 
   enum BatchFactory implements Factory {
-    INSTANCE;
-
-    @SuppressWarnings("Immutable") // observably immutable
-    private final LoadingCache<JobInfo, BatchFlinkExecutableStageContext> cachedContexts;
-
-    BatchFactory() {
-      cachedContexts =
-          CacheBuilder.newBuilder()
-              .weakValues()
-              .build(
-                  new CacheLoader<JobInfo, BatchFlinkExecutableStageContext>() {
-                    @Override
-                    public BatchFlinkExecutableStageContext load(JobInfo jobInfo) throws Exception {
-                      return create(jobInfo);
+    REFERENCE_COUNTING {
+      private static final int TTL_IN_SECONDS = 30;
+      int MAX_RETRY = 3;
+
+      @SuppressWarnings("Immutable") // observably immutable
+      private final ScheduledExecutorService executor =
+          Executors.newScheduledThreadPool(1, new ThreadFactoryBuilder().setDaemon(true).build());
+
+      @SuppressWarnings("Immutable") // observably immutable
+      private final ConcurrentHashMap<String, ContextWrapper> keyRegistry =
+          new ConcurrentHashMap<>();
+
+      @Override
+      public FlinkExecutableStageContext get(JobInfo jobInfo) {
```

Comment:
The caching behavior is complex enough that I think it's worth making this its own top-level class and adding tests. If it's not too much work to do so, it may also be worth making generic.

Please note my comment below about switching to a `close()` method on the Context rather than `release(jobInfo)`.